spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java

```diff
@@ -0,0 +1,161 @@
+				}
+				else {
+					throw new IllegalStateException(
+							"A default binder has been requested, but there is more than one binder available: "
+									+ StringUtils.collectionToCommaDelimitedString(binderConfigurations.keySet()) + ", and"
+									+ " no default binder has been set.");
+				}
+			}
+		} else {
+			configurationName = name;
+		}
+		if (!binderInstanceCache.containsKey(configurationName)) {
+			BinderConfiguration binderConfiguration = binderConfigurations.get(configurationName);
+			if (binderConfiguration == null) {
+				throw new IllegalStateException("Unknown binder configuration: " + configurationName);
+			}
+			Properties binderProperties = binderConfiguration.getProperties();
+			// Convert all properties to arguments, so that they receive maximum precedence
+			ArrayList<String> args = new ArrayList<>();
+			for (Map.Entry<Object, Object> property : binderProperties.entrySet()) {
+				args.add(String.format("--%s=%s",property.getKey(),property.getValue()));
+			}
+			// Initialize the domain with a unique name based on the bootstrapping context setting
+			String defaultDomain = environment != null ? environment.getProperty("spring.jmx.default-domain") : null;
+			if (defaultDomain == null) {
+				defaultDomain = "";
+			}
+			else {
+				defaultDomain += ".";
+			}
+			args.add("--spring.jmx.default-domain=" + defaultDomain + "binder." + configurationName);
+			SpringApplicationBuilder springApplicationBuilder =
+					new SpringApplicationBuilder()
+							.sources(binderConfiguration.getBinderType().getConfigurationClasses())
+							.sources(SeedConfiguration.class)
+							.bannerMode(Mode.OFF)
+							.web(false);
+			ConfigurableApplicationContext binderProducingContext =
+					springApplicationBuilder.run(args.toArray(new String[args.size()]));
+			@SuppressWarnings("unchecked")
+			Binder<T> binder = (Binder<T>) binderProducingContext.getBean(Binder.class);
+			binderInstanceCache.put(configurationName, new BinderInstanceHolder<>(binder, binderProducingContext));
+		}
+		return binderInstanceCache.get(configurationName).getBinderInstance();
+	}
+
+	/**
+	 * Utility class for storing {@link Binder} instances, along with their associated contexts.
+	 *
+	 * @param <T>
+	 */
+	private static class BinderInstanceHolder<T> {
+
+		private final Binder<T> binderInstance;
+
+		private final ConfigurableApplicationContext binderContext;
+
+		public BinderInstanceHolder(Binder<T> binderInstance, ConfigurableApplicationContext binderContext) {
+			this.binderInstance = binderInstance;
+			this.binderContext = binderContext;
+		}
+
+		public Binder<T> getBinderInstance() {
+			return binderInstance;
+		}
+
+		public ConfigurableApplicationContext getBinderContext() {
+			return binderContext;
+		}
+	}
+
+	/**
+	 * Configuration class that enables autoconfiguration for the binders
+	 */
+	// TODO: Reconsider the use of autoconfiguration as part of binder configuration refactoring
+	@EnableAutoConfiguration
```

Comment:
Shouldn't the `enable` auto configuration be part of the actual binder `configuration` classes defined from the binder implementors?